src/core/ChannelsProvider.tsx

```diff
@@ -0,0 +1,90 @@
+import { Channel, PresenceChannel } from "pusher-js";
+import React, { useCallback, useRef } from "react";
+import { ChannelsContextValues } from "./types";
+
+import { usePusher } from "./usePusher";
+
+// context setup
+const ChannelsContext = React.createContext<ChannelsContextValues>({});
+export const __ChannelsContext = ChannelsContext;
+
+type AcceptedChannels = Channel | PresenceChannel;
+type ConnectedChannels = {
+  [channelName: string]: AcceptedChannels[];
+};
+
+/**
+ * Provider that creates your channels instances and provides it to child hooks throughout your app.
+ */
+
+export const ChannelsProvider: React.FC<{ children: React.ReactNode }> = ({
+  children,
+}) => {
+  const { client } = usePusher();
+  const connectedChannels = useRef<ConnectedChannels>({});
+
+  const subscribe = useCallback(
+    <T extends Channel & PresenceChannel>(channelName: string) => {
+      /** Return early if there's no client */
+      if (!client) return;
+
+      /** Return early if channel name is falsy */
+      if (!channelName) return;
+
+      /** Subscribe to channel and set it in state */
+      const pusherChannel = client.subscribe(channelName);
+      connectedChannels.current[channelName] = [
+        ...(connectedChannels.current[channelName] || []),
+        pusherChannel,
+      ];
+      return pusherChannel as T;
+    },
+    [client, connectedChannels]
+  );
+
+  const unsubscribe = useCallback(
+    (channelName: string) => {
+      /** Return early if there's no client */
+      if (!client) return;
+      /** Return early if channel name is falsy */
+      if (!channelName) return;
+      /** If no connection, just skip*/
+      if (!connectedChannels.current[channelName]?.length) return;
+
+      /** If just one connection, unsubscribe totally*/
+      if (connectedChannels.current[channelName].length === 1) {
+        client.unsubscribe(channelName);
+        delete connectedChannels.current[channelName];
+      } else {
+        connectedChannels.current[channelName].pop();
+      }
+    },
+    [connectedChannels, client]
+  );
+
+  const getChannel = useCallback(
+    <T extends Channel & PresenceChannel>(channelName: string) => {
+      /** Return early if there's no client */
+      if (!client) return;
+      /** Return early if channel name is falsy */
+      if (!channelName) return;
+      /** Return early if channel is not in state */
+      if (!connectedChannels.current[channelName]) return;
+      /** Return channel */
+      return connectedChannels.current[channelName][0] as T;
+    },
+    [connectedChannels, client]
```

Comment:
probably don't need `connectedChannels` in dependencies as it's a ref. Why did you opt for a ref rather than `useState`? Wouldn't `useState` refresh all the subscribers to get the updated channels?

Reply:
I've tried with useState but subscribe and unsubscribe reference changes every time `connectedChannels` changes, this makes `useChannel` `useEffect` to runs infinitely, tried with useCallbacks, useMemos, even using the callback of the setState instead of using the value of the state. Found the most stable option using the ref.

Probably I'm missing something, but couldn't get it work with useState :(

Reply:
Also, with useState if some parent context changes, will lose all the connections right?